[clang] Implement pragma clang section on COFF targets

vtz opened the pull request:
This patch implements the directive pragma clang section on COFF targets with the exact same features available on ELF and Mach-O.

vtz commented:
Dear @rnk, I believe you're the right one to review this PR. If not, could you please share the appropriate reviewer?
Thank you very much!

DavidSpickett commented on clang/test/Sema/pragma-clang-section-coff.c:
```diff
@@ -0,0 +1,7 @@
+// expected-no-diagnostics
+#pragma clang section bss = "mybss.1" data = "mydata.1" rodata = "myrodata.1" text = "mytext.1"
+#pragma clang section bss="" data="" rodata="" text=""
+#pragma clang section
```
I just compared with `clang/test/Sema/pragma-clang-section-macho.c`, that has some invalid examples in it. Do you have any invalid cases for coff?

vtz replied:
Solved with an aditional RUN command in the existing test

DavidSpickett commented on llvm/lib/CodeGen/TargetLoweringObjectFileImpl.cpp:
```diff
@@ -1677,6 +1677,22 @@ MCSection *TargetLoweringObjectFileCOFF::getExplicitSectionGlobal(
+      Name = Attrs.getAttribute("data-section").getValueAsString();
+    }
+  }
+
```
I'm not sure if it's required / possible / useful to test this in llvm as well, but if you haven't looked into it already, you might want to look at the commits that added the Macho and ELF versions and see if there are llvm tests there too.

vtz replied:
I tried making a new test based on llvm/test/CodeGen/ARM/clang-section.ll, but it ended up equal so I added an extra RUN command to this test as well

DavidSpickett commented on llvm/lib/CodeGen/TargetLoweringObjectFileImpl.cpp:
```diff
@@ -1677,6 +1677,22 @@ MCSection *TargetLoweringObjectFileCOFF::getExplicitSectionGlobal(
+  const GlobalVariable *GV = dyn_cast<GlobalVariable>(GO);
+  if (GV && GV->hasImplicitSection()) {
+    auto Attrs = GV->getAttributes();
+    if (Attrs.hasAttribute("bss-section") && Kind.isBSS()) {
```
I think https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements applies here.

Despite clang-format deciding to put one of the conditions onto 2 lines.

vtz replied:
Done

DavidSpickett commented on llvm/lib/CodeGen/TargetLoweringObjectFileImpl.cpp:
```diff
@@ -1677,6 +1677,22 @@ MCSection *TargetLoweringObjectFileCOFF::getExplicitSectionGlobal(
+      Name = Attrs.getAttribute("relro-section").getValueAsString();
+    } else if (Attrs.hasAttribute("data-section") && Kind.isData()) {
+      Name = Attrs.getAttribute("data-section").getValueAsString();
+    }
```
I guess that the conditions here list all the valid combinations. If we got to the implicit `else`, would falling through without setting name be ok?

It would be as if `if (GV && GV->hasImplicitSection()) {` was false, so I think it would do something valid, since that path already exists.

vtz replied:
"would falling through without setting name be ok?" yes, in this case we would keep the section's original name, I don't see the need for an explicit else here.

DavidSpickett commented on llvm/lib/CodeGen/TargetLoweringObjectFileImpl.cpp:
```diff
@@ -1677,6 +1677,22 @@ MCSection *TargetLoweringObjectFileCOFF::getExplicitSectionGlobal(
+
+  const GlobalVariable *GV = dyn_cast<GlobalVariable>(GO);
+  if (GV && GV->hasImplicitSection()) {
+    auto Attrs = GV->getAttributes();
```
You could consider making this a static function to be used by both places. Especially if there is something like this already for the other formats.

vtz replied:
Done

DavidSpickett reviewed (commented):
I'm not an expert here but added some generic comments for now. FYI with the llvm conference last week, many people are still away so review pace is slower right now.

Since this is a Clang change it should have a release note, I think it can go in https://clang.llvm.org/docs/ReleaseNotes.html#windows-support.

The source of that file is `clang/docs/ReleaseNotes.rst`. You can use the `#GH` syntax to link back to this pull request. You'll see other examples in the doc.

(RST can be a bit strange if you are used to Markdown so if in doubt, emulate what's already there)

DavidSpickett commented:
Also I edited the title tags. They aren't an exact science but clang seems the most appropriate here.

efriedma-quic commented:
(Not sure who the right reviewers are for XCOFF; I tried to guess.)

hubert-reinterpretcast reviewed (commented):
The documentation (see https://clang.llvm.org/docs/LanguageExtensions.html) needs to be updated. It currently says that "This feature is only defined to work sensibly for ELF and Mach-O targets".

hubert-reinterpretcast commented on llvm/lib/CodeGen/TargetLoweringObjectFileImpl.cpp:
```diff
@@ -2378,13 +2394,28 @@ MCSection *TargetLoweringObjectFileXCOFF::getExplicitSectionGlobal(
   StringRef SectionName = GO->getSection();
```
See `report_fatal_error` above.
I think the new code should be moved up so that `SectionName` is correct by the time we reach the next line.

I also believe that this `pragma` should suppress the application of `-mtocdata`.
For explicit section attributes and `-mtocdata=<var>`, there is a warning implemented:
```
-mtocdata option is ignored for <var> because variable has a section attribute
```

vtz replied:
The focus was from the beginning on COFF format. We misinterpreted and thought the XCOFF would be the same way. It doesn't look to be so, so we decided to drop the changes for XCOFF from this PR. This also makes this PR more isolated.

hubert-reinterpretcast reviewed (commented):
There doesn't seem to be changes to handle the text section cases?

rnk reviewed (commented):
This is an LLVM code change, not a clang code change. It's an important principle that we test LLVM at the smallest reasonable granularity. Can you replace the clang test with an IR test? I'm sure we already have existing IR carrying existing sections, we just need to test it with a new target.

vtz commented:
> This is an LLVM code change, not a clang code change. It's an important principle that we test LLVM at the smallest reasonable granularity. Can you replace the clang test with an IR test? I'm sure we already have existing IR carrying existing sections, we just need to test it with a new target.

@rnk this is also addressed with our new commit.

Please refer to my comment above and quoted below:
> I tried making a new test based on llvm/test/CodeGen/ARM/clang-section.ll, but it ended up equal so I added an extra RUN command to this test as well

DavidSpickett commented:
My comments have been addressed.

vtz committed:
[llvm] Implement pragma clang section on COFF targets

This patch implements the directive pragma clang section on
    COFF targets with the exact same features available on ELF
    and Mach-O.

efriedma-quic commented on llvm/test/CodeGen/ARM/clang-section.ll:
```diff
@@ -1,4 +1,5 @@
 ;RUN: llc -mtriple=armv7-eabi %s -o - | FileCheck %s
+;RUN: llc -mtriple=armv7-msvc %s -o - | FileCheck %s
```
"armv7-msvc" doesn't actually trigger COFF output.  Try "armv7-windows-msvc".

vtz replied:
Changed to x86. I hope this is fine.

efriedma-quic replied:
That's fine.  But the CHECK lines in the new test don't look right (`nobits` etc. are ELF-specific).

vtz replied:
You're right. Improved it.

efriedma-quic commented on llvm/lib/CodeGen/TargetLoweringObjectFileImpl.cpp:
```diff
@@ -1684,6 +1676,7 @@ MCSection *TargetLoweringObjectFileCOFF::getExplicitSectionGlobal(
       Name == getInstrProfSectionName(IPSK_covname, Triple::COFF,
                                       /*AddSegmentInfo=*/false))
     Kind = SectionKind::getMetadata();
+
```
Unnecessary change.

vtz replied:
Removed.

efriedma-quic commented:
(The PR was automatically closed because vtz:coff-pragma-section doesn't have any changes anymore.  It should be possible to reopen if you push the correct commit to the branch.)

vtz commented:
@efriedma-quic , I guess all your comments were addressed.

efriedma-quic reviewed (approved):
LGTM

efriedma-quic reviewed (approved):
LGTM
